allennlp/models/language_model.py

```diff
@@ -0,0 +1,206 @@
+from typing import Optional, Dict
+
+from overrides import overrides
+import torch
+import torch.nn as nn
+
+from allennlp.common import Params
+from allennlp.data import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules import TextFieldEmbedder
+from allennlp.common.checks import check_dimensions_match
+from allennlp.nn import InitializerApplicator, RegularizerApplicator
+
+from allennlp.modules.lm_rnn import LMRNN
+from allennlp.modules.adaptive import AdaptiveSoftmax
+from allennlp.training.metrics.perplexity import Perplexity
+
+@Model.register("WordLM")
+class WordLM(Model):
+    """
+    This ``WordLM`` is a word level language model.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        A Vocabulary, required in order to compute sizes for input/output projections.
+    text_field_embedder : ``TextFieldEmbedder``, required
+        Used to embed the ``tokens`` ``TextField`` we get as input to the model.
+    rnn : ``LMRNN``, required
+        The Recurrent Neural Networks for language modeling.
+    softmax : ``AdaptiveSoftmax``
```

Comment:
Same here with the softmax -- we should allow any module that computes the softmax including full softmax, sampled, etc.

Reply:
i'll create a registrable class ```softmax_with_nll_loss`` for this